### ISSUE TITLE
Search in settings

### ISSUE THREAD

**Issue by Ahmadhosseinbor** [osmandapp/OsmAnd#13973]
Ability to search settings to find features faster

**Comment by franco999** [comment]
The large amount of settings and customization is one of the advantages that OsmAnd has over other similar applications, but this is overshadowed by itself to an average user, for the same reason. The truth is that it would be very helpful and practical to have this function in the style of Android Settings.

**Comment by mariush444** [comment]
@vshcherb 
Is it possible to move this issue to 4.7 milestone?
4.3/4.4 is puvlished long time ago but this issue was forgotten sonehow.

**Comment by KnollFrank** [comment]
Do you mean these settings?

<table style="width:100%">
  <tr>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/23587457/515dbe90-1633-4089-86ce-60ff03944fd7"  width="300" height="auto"/> </th>
  </tr>
</table>

**Comment by KnollFrank** [comment]
Please see the pull request https://github.com/osmandapp/OsmAnd/pull/20016

**Comment by KnollFrank** [comment]
https://github.com/osmandapp/OsmAnd/assets/23587457/aba6ea11-1ff7-4973-b2ee-2da9caa7f44e

**Comment by franco999** [comment]
> Search.in.settings.mp4

At first glance, I like it and it would be an excellent feature for novice or clueless users.

**Comment by franco999** [comment]
Sorry to insist, but is there any news on this? Any prototype merged into main project for testing?

**Comment by vshcherb** [comment]
I guess work has been done in the library which is good and bad, so there will be much less code to review / maintain by the team. Though it requires more work, probably and definitely some API will need to be configured to customize UI for specific needs and integrate.

**Comment by KnollFrank** [comment]
Should the preference dialogs be searchable, e.g. those dialogs handled in PreferenceFragmentCompat.onDisplayPreferenceDialog()?

**Comment by KnollFrank** [comment]
Is this feature still needed?

**Comment by franco999** [comment]
I don't know how many more hours of work it takes and how complicated the code might be, but I'm still looking forward to this feature and still think it can simplify the search for options, both for newbies and experienced users.

**Comment by mariush444** [comment]
> Should the contents of preference dialogs be searchable, e.g. those dialogs displayed in PreferenceFragmentCompat.onDisplayPreferenceDialog()?
> 
> For example, if you type "cone-shaped" in the search field, then should the preference named "view angle" (which doesn't have the string "cone-shaped" in its title nor in its description) be in the search results, because this preference will open a dialog containing "cone-shaped" in its description when you click it?

I expect searching existing text only. I do not expect any prediction, guessing or AI.
But this is my own point of view.

**Comment by franco999** [comment]
I'll settle for looking at titles and descriptions. Something basic for orientation.

**Comment by mnalis** [comment]
> Is this feature still needed?

@KnollFrank If you are talking about _"Search in settings"_ generally - absolutely **yes**: for me that is one of the most eagerly awaited features in OsmAnd for quite some time! I see in that linked PR that it is _a lot_ of work (quite more than I would've anticipated) but know that your effort is _**really**_ appreciated! Please persevere! :heart: 

And if you're talking only about that last comment from September:

> For example, if you type "cone-shaped" in the search field, then should the preference named "view angle" (which doesn't have the string "cone-shaped" in its title nor in its description) be in the search results

I guess it would be quite nice, but if that is too much work, just searching over title+description would already be an enormous improvement. 

OsmAnd is very powerful (esp. compared to apps like OrganicMaps), but that power comes at a price; and complexity of finding a correct setting is one of the biggest parts of that price. 

Just like with Android system settings: those days I find things there in 95%+ of cases just by searching instead of trying to navigate that menu-submenu-submenu-threedots-submenu-submenu maze... It horrifies me even to think of the possibility that I would need to get back to manual navigation there. 
I'm expecting similar relief when preferences search is implemented in OsmAnd, and can't wait for that!

**Comment by KnollFrank** [comment]
Hello, perhaps someone could test the current status of the settings search and suggest changes that still need to be implemented?

**Comment by franco999** [comment]
> Hello, perhaps someone could test the current status of the settings search and suggest changes that need to be implemented?

I would love to be a beta tester as an end user, as I cannot help as a Java programmer. 
How would this be implemented, from the nightly?

**Comment by KnollFrank** [comment]
> How would this be implemented, from the nightly?

Thanks for testing, but I don't know how to integrate Settings Search in a nightly build. Hopefully someone can help.

**Comment by mariush444** [comment]
today nightly 05.01.25 doesn't contain search menu.
Maybe I don't know where to find it. Can you show screenshot?

**Comment by KnollFrank** [comment]
> today nightly 05.01.25 doesn't contain search menu.
> Maybe I don't know where to find it. Can you show screenshot?

I think the code implementing Settings Search first needs to be merged by someone who has the rights to do it. Once this is done then you can start searching by clicking the burger menu then clicking "Settings" and then clicking the magnifying glass symbol in the upper right corner like in the Video above.

**Comment by mariush444** [comment]
probably @vshcherb or @RZR-UA can help you to merge your code

**Comment by KnollFrank** [comment]
Hi,

[here](https://github.com/KnollFrank/OsmAnd/releases/tag/testing) you can download the file [OsmAnd-nightlyFree-legacy-fat-debug.apk](https://github.com/KnollFrank/OsmAnd/releases/download/testing/OsmAnd-nightlyFree-legacy-fat-debug.apk) for installing on a device and start testing.

Settings search can be started by
1. clicking on the burger menu then
2. clicking "Settings" then
3. clicking the magnifying glass symbol in the upper right corner and then
4. searching for some settings by typing some keywords in the text field.

Cheers, Frank

**Comment by franco999** [comment]
I haven't tried it thoroughly yet, but I like it. I find it a bit difficult to understand the target option because of the number of profiles there are. 

I will continue testing and share my experience.

![Screenshot_20250107-192943](https://github.com/user-attachments/assets/a3f57a6d-7393-469c-92be-661d2ef65da9)
![Screenshot_20250107-192553](https://github.com/user-attachments/assets/1e007023-8d43-4e48-89ee-fe133ec2f802)

**Comment by KnollFrank** [comment]
> I find it a bit difficult to understand the target option because of the number of profiles there are. 

Maybe the search results only differing in the profile part should be grouped under a single item which reveals the other search results when clicked?

**Comment by mariush444** [comment]
I think it is OK like this.
If we implement grouping, sorting, filtering it be to sofisticated.
Searching should as easy as possible because menu is too complicated.
That's my opinion.

**Comment by franco999** [comment]
> > I find it a bit difficult to understand the target option because of the number of profiles there are.
> 
> Maybe the search results only differing in the profile part (e.g. all 7 search results for "condu" in your example above) should be grouped under a single item which reveals all search results of this group when clicked?

For example, there are profiles that I have deactivated and I don't use them at all, therefore, I am not interested in showing them in the search, since it makes it more complicated to find the desired option. 

Could it be that it only searches within the profiles that are activated? This would reduce a bit the redundancies of options.

**Comment by KnollFrank** [comment]
> Could it be that it only searches within the profiles that are activated?

Ok, I will try to Implement this next.

**Comment by KnollFrank** [comment]
> Could it be that it only searches within the profiles that are activated? This would reduce a bit the redundancies of options.

Ok, the search is restricted to activated profiles: [OsmAnd-nightlyFree-legacy-fat-debug.apk](https://github.com/KnollFrank/OsmAnd/releases/download/testing/OsmAnd-nightlyFree-legacy-fat-debug.apk)

**Comment by franco999** [comment]
> > Could it be that it only searches within the profiles that are activated? This would reduce a bit the redundancies of options.
> 
> Ok, the search is restricted to activated profiles: [OsmAnd-nightlyFree-legacy-fat-debug.apk](https://github.com/KnollFrank/OsmAnd/releases/download/testing/OsmAnd-nightlyFree-legacy-fat-debug.apk)

I like it, it has greatly improved the redundancies of profiles that are not of interest. For me it is excellent.

Now, I think. If I want to search for an option within a specific profile, it would work to do for example "driving, speed camera" to show the text "*speed camera*" only from the "Driving" profile or is that asking too much. That is, use the advanced search to filter better.

What do others think?

**Comment by mariush444** [comment]
in my opinion, go to `driving` profile then search `speed camera`

**Comment by mnalis** [comment]
> in my opinion, go to driving profile then search speed camera

I don't know about that; if I know where I need to go, I don't really need search functionality, I can just go there...

I'm OK with options only being found in enabled profiles (and enabled plugins), if we want to keep it simple. So enable car profile if you want it to be searched too. 

---

An alternative which would be great however (but I'm not sure how much work that is) is to also have a single checkbox "search all" in Search, which (if enabled) would _also_ search inside disabled profiles (and - if possible - in installed but disabled plugins too).
Or, alternatively, always _only_ search in enabled profiles/plugins, but at the end of search results have button "Search in disabled profiles/plugins" which (if pressed) would search there too.

**Comment by mnalis** [comment]
> Ok, the search is now restricted to activated profiles: [OsmAnd-nightlyFree-legacy-fat-debug.apk](https://github.com/KnollFrank/OsmAnd/releases/download/testing/OsmAnd-nightlyFree-legacy-fat-debug.apk)

Thanks!  By the way, feel free to leave previous versions also published under releases too with different descriptions; that way people can try different versions and see which one works better for them. I.e. only add new (pre-)releases, do not remove old ones..

**Comment by KnollFrank** [comment]
> ... to also have a single checkbox "search all"

Ok, I will try to implement this feature.

**Comment by KnollFrank** [comment]
Here is a new version with a checkbox "search inside disabled profiles" on the search page: https://github.com/KnollFrank/OsmAnd/releases/tag/testing_v2

**Comment by KnollFrank** [comment]
> ... to also have a single checkbox "search all" in Search, which (if enabled) would _also_ search inside disabled profiles (and - if possible - in installed but disabled plugins too).

I don't know how to handle search in **disabled** plugins, because when the user clicks a search result pointing to a setting inside a disabled plugin then the yet disabled plugin would have to automatically be enabled in order to navigate to the search result?

**Comment by franco999** [comment]
Just to note, I tried to do a test search for the `Text size` option and others found in “**Configure map**”, but nothing was found. Is this the case or is it a “bug”?
<img alt="text size" height="300" src="https://github.com/user-attachments/assets/b0ee3f7e-1cfd-4c0a-bcfe-eb70d44c936e" />

 I also tried searching for `Speedometer` but it does not lead directly to such an option, but goes all the way to the “**Configure screen**” option within the profile.
<img alt="configure screen" height="300" src="https://github.com/user-attachments/assets/2111866a-7694-42a2-a898-90598e323970" />

**Comment by KnollFrank** [comment]
> Just to note, I tried to do a test search for the `Text size` option and others found in “**Configure map**”, but nothing was found. Is this the case or is it a “bug”?

The “**Configure map**” preference navigates to a new `android.content.Intent` when clicked which I don't know how to handle yet when building the search database. I will try to implement this feature.
 
> I also tried searching for `Speedometer` but it does not lead directly to such an option, but goes all the way to the “**Configure screen**” option within the profile.

Yes, the search library currently only navigates to the preference “**Configure screen**” but doesn't click it in order to open the fragment or dialog containing the `Speedometer` item. I will try to implement this missing feature next.

**Comment by franco999** [comment]
If it is complicated to implement, in my opinion, there is no problem. I was just making observations that made me doubt. 
But you're doing great with this feature. I love how it's progressing and you're an amazing developer.

**Comment by mnalis** [comment]
> I don't know how to handle search in disabled plugins, because when the user clicks a search result pointing to a setting inside a disabled plugin then the yet disabled plugin would have to automatically be enabled in order to navigate to the search result?

Ah, yes, I didn't think of that. Agreed, it is probably best to just ignore disabled plugins then; if people want to search them, they'd need to enable them first - sounds reasonable to me! (auto-enabling them sounds like a bad idea)

**Comment by KnollFrank** [comment]
> I also tried searching for `Speedometer` but it does not lead directly to such an option, but goes all the way to the “**Configure screen**” option within the profile.

This version opens the dialog containing the item "Speedometer: https://github.com/KnollFrank/OsmAnd/releases/tag/testing_v3

**Comment by KnollFrank** [comment]
> Just to note, I tried to do a test search for the `Text size` option and others found in “**Configure map**”, but nothing was found.

The `Text size` preference is implemented using the custom `net.osmand.plus.widgets.ctxmenu.data.ContextMenuItem` class which is not part of the standard android `androidx.preference.Preference` class hierarchy. These preferences can't be handled by the settings search library right now. Is it possible to refactor these preferences to be implemented as `androidx.preference.Preference`s or should the search library be enhanced to handle custom preferences?

Cheers, Frank

**Comment by KnollFrank** [comment]
Hi @franco999, 

> Just to note, I tried to do a test search for the `Text size` option and others found in “**Configure map**”, but nothing was found. Is this the case or is it a “bug”?

Some settings of  “**Configure map**” are searchable now in this version: https://github.com/KnollFrank/OsmAnd/releases/tag/testing_v4

**Comment by franco999** [comment]
> Hi [@franco999](https://github.com/franco999),
> 
> > Just to note, I tried to do a test search for the `Text size` option and others found in “**Configure map**”, but nothing was found. Is this the case or is it a “bug”?
> 
> Some settings of “**Configure map**” are searchable now in this version: https://github.com/KnollFrank/OsmAnd/releases/tag/testing_v4

When clicking on the magnifying glass and loading search data, the app crashed and closed unexpectedly.

Device : devon
Brand : motorola
Manufacturer : motorola
Model : moto g32
Product : devon_g
Build : T2SNS33.73-22-3-13
Version : 13
App Version : OsmAnd Nightly
Apk Version : 5.0.0 5000

[exception (1).log](https://github.com/user-attachments/files/19618719/exception.1.log)

**Comment by KnollFrank** [comment]
> > Hi [@franco999](https://github.com/franco999),
> > > Just to note, I tried to do a test search for the `Text size` option and others found in “**Configure map**”, but nothing was found. Is this the case or is it a “bug”?
> > 
> > 
> > Some settings of “**Configure map**” are searchable now in this version: https://github.com/KnollFrank/OsmAnd/releases/tag/testing_v4
> 
> When clicking on the magnifying glass and loading search data, the app crashed and closed unexpectedly.
> 
> Device : devon Brand : motorola Manufacturer : motorola Model : moto g32 Product : devon_g Build : T2SNS33.73-22-3-13 Version : 13 App Version : OsmAnd Nightly Apk Version : 5.0.0 5000
> 
> [exception (1).log](https://github.com/user-attachments/files/19618719/exception.1.log)

Yes, I sometimes have the same error, but don't know how to reproduce it reliably. Could you please try again (maybe delete OsmAnd completely before installing this version?).

**Comment by KnollFrank** [comment]
What are the requirements for this feature to be considered implemented?

**Comment by KnollFrank** [comment]
> > > Hi [@franco999](https://github.com/franco999),
> > > > Just to note, I tried to do a test search for the `Text size` option and others found in “**Configure map**”, but nothing was found. Is this the case or is it a “bug”?
> > > 
> > > 
> > > Some settings of “**Configure map**” are searchable now in this version: https://github.com/KnollFrank/OsmAnd/releases/tag/testing_v4
> > 
> > 
> > When clicking on the magnifying glass and loading search data, the app crashed and closed unexpectedly.
> > Device : devon Brand : motorola Manufacturer : motorola Model : moto g32 Product : devon_g Build : T2SNS33.73-22-3-13 Version : 13 App Version : OsmAnd Nightly Apk Version : 5.0.0 5000
> > [exception (1).log](https://github.com/user-attachments/files/19618719/exception.1.log)
> 
> Yes, I sometimes have the same error, but don't know how to reproduce it reliably. Could you please try again (maybe delete OsmAnd completely before installing this version?).



> > Hi [@franco999](https://github.com/franco999),
> > > Just to note, I tried to do a test search for the `Text size` option and others found in “**Configure map**”, but nothing was found. Is this the case or is it a “bug”?
> > 
> > 
> > Some settings of “**Configure map**” are searchable now in this version: https://github.com/KnollFrank/OsmAnd/releases/tag/testing_v4
> 
> When clicking on the magnifying glass and loading search data, the app crashed and closed unexpectedly.
> 
> Device : devon Brand : motorola Manufacturer : motorola Model : moto g32 Product : devon_g Build : T2SNS33.73-22-3-13 Version : 13 App Version : OsmAnd Nightly Apk Version : 5.0.0 5000
> 
> [exception (1).log](https://github.com/user-attachments/files/19618719/exception.1.log)

Sorry, can you confirm that the program doesn't crash on a second launch? I know this isn't a solution yet.

**Comment by KnollFrank** [comment]
Yes, @franco999 it looks misaligned, but I have no motorola and can't reproduce.
